### PR TITLE
Hotfix/libp2p details

### PR DESCRIPTION
**Pull request by cmr**
First round of small libp2p-related fixes.

### CODE REVIEW

**Comment by jkrauska** [comment]
key gen works

```
$ coda advanced  generate-libp2p-keypair
libp2p keypair:
23jhTayhW...
```

libp2p starts

```
2019-08-30 01:35:22 UTC [Info] libp2p peer ID this session is $peer_id
	peer_id: "12D3KooWHsWVd62ezNCrwj8dGrmswmizAVVfc9yqb1dMqqEWa861"**
```

still crashes
```
Generating crash report
	exn: "\"RPC #1 failed: \\\"internal RPC error error: helper not yet configured\\\"\""
```

**Comment by jkrauska** [comment]
farther

```
\":\"/home/ubuntu/.coda-config/coda_net2/libp2p_helper\",\"ifaces\":[],\"network_id\":\"libp2p phase2 test network\"}}"
2019-08-30 20:05:57 UTC [Debug] log message from libp2p_helper: $line
        line: "{\"id\":1,\"level\":4,\"message\":\"All 0 tables opened in 0s\\n\",\"module\":\"badger\",\"time\":\"2019-08-30T20:05:57.292296129Z\"}"
2019-08-30 20:05:57 UTC [Debug] log message from libp2p_helper: ""
2019-08-30 20:05:57 UTC [Debug] log message from libp2p_helper: $line
        line: "{\"id\":2,\"level\":4,\"message\":\"All 0 tables opened in 0s\\n\",\"module\":\"badger\",\"time\":\"2019-08-30T20:05:57.298912181Z\"}"
2019-08-30 20:05:57 UTC [Debug] log message from libp2p_helper: ""
2019-08-30 20:05:57 UTC [Debug] log message from libp2p_helper: $line
        line: "{\"id\":3,\"level\":2,\"message\":\"failed to find good external addr from peerhost\",\"module\":\"mdns\",\"time\":\"2019-08-30T20:05:57.304705351Z\"}"
2019-08-30 20:05:57 UTC [Debug] log message from libp2p_helper: ""
2019-08-30 20:05:57 UTC [Debug] log message from libp2p_helper: $line
        line: "2019/08/30 20:05:57 Announcing ourselves for /coda/0.0.1/libp2p phase2 test network"
2019-08-30 20:05:57 UTC [Trace] handling line from helper: "{\"seqno\":1,\"success\":\"configure success\"}"
2019-08-30 20:05:57 UTC [Trace] sending line to libp2p_helper: $line
        line: "{\"seqno\":2,\"method\":\"addStreamHandler\",\"body\":{\"protocol\":\"coda/0.0.1/discovery-port\"}}"
2019-08-30 20:05:57 UTC [Trace] handling line from helper: $line
        line: "{\"seqno\":2,\"success\":\"addStreamHandler success\"}"
2019-08-30 20:05:57 UTC [Debug] log message from libp2p_helper: $line
        line: "{\"id\":4,\"level\":4,\"message\":\"starting bootstrap query (0/1) to random ID (routing table size was 0)\",\"module\":\"dht\",\"time\":\"2019-08-30T20:05:57.305811824Z\"}"
2019-08-30 20:05:57 UTC [Trace] sending line to libp2p_helper: $line
        line: "{\"seqno\":3,\"method\":\"listen\",\"body\":{\"iface\":\"/ip4/0.0.0.0/tcp/28675\"}}"
2019-08-30 20:05:57 UTC [Debug] log message from libp2p_helper: ""
2019-08-30 20:05:57 UTC [Debug] log message from libp2p_helper: $line
        line: "{\"id\":5,\"level\":4,\"message\":\"finished bootstrap query (0/1) to random ID (routing table size is now 0)\",\"module\":\"dht\",\"time\":\"2019-08-30T20:05:57.306393048Z\"}"
2019-08-30 20:05:57 UTC [Debug] log message from libp2p_helper: ""
2019-08-30 20:05:57 UTC [Debug] log message from libp2p_helper: $line
        line: "{\"id\":6,\"level\":2,\"message\":\"error bootstrapping: failed to find any peer in table\",\"module\":\"dht\",\"time\":\"2019-08-30T20:05:57.306469242Z\"}"
2019-08-30 20:05:57 UTC [Debug] log message from libp2p_helper: ""
2019-08-30 20:05:57 UTC [Trace] handling line from helper: $line
        line: "{\"seqno\":3,\"success\":[\"/ip4/127.0.0.1/tcp/28675\",\"/ip4/172.31.1.139/tcp/28675\"]}"
2019-08-30 20:05:57 UTC [Trace] sending line to libp2p_helper: $line
        line: "{\"seqno\":4,\"method\":\"beginAdvertising\",\"body\":{}}"
2019-08-30 20:05:57 UTC [Fatal] Unhandled top-level exception: "Not_found"
Generating crash report
2019-08-30 20:05:57 UTC [Debug] Successfully wrote the visualization of the registered masks at location: /home/ubuntu/.coda-config/coda_crash_report_2019-08-30_20-05-57.338841/registered_masks.dot.
2019-08-30 20:05:57 UTC [Trace] Shutdown before Coda instance was created, not saving a visualization


  ☠  Coda daemon crashed. The Coda Protocol developers would like to know why!

  Please:
    Open an issue:
      <https://github.com/CodaProtocol/coda/issues/new>

    Briefly describe what you were doing and attach the crash report /home/ubuntu/.coda-config/coda_crash_report_2019-08-30_20-05-57.338841.tar.gz
```

**Comment by jkrauska** [comment]
@cmr 

FWIW 172.31.1.139 is the private IP of the host.

External IP is 52.42.148.210

Should I add -external IP?

**Comment by jkrauska** [comment]
I added -external-ip $myip, but the libp2p layer didn't see or use that.

same 172 stuff

```
2019-08-30 20:11:25 UTC [Trace] handling line from helper: $line
        line: "{\"seqno\":2,\"success\":\"addStreamHandler success\"}"
2019-08-30 20:11:25 UTC [Trace] sending line to libp2p_helper: $line
        line: "{\"seqno\":3,\"method\":\"listen\",\"body\":{\"iface\":\"/ip4/0.0.0.0/tcp/28675\"}}"
2019-08-30 20:11:25 UTC [Trace] handling line from helper: $line
        line: "{\"seqno\":3,\"success\":[\"/ip4/127.0.0.1/tcp/28675\",\"/ip4/172.31.1.139/tcp/28675\"]}"
```

**Comment by cmr** [comment]
Update: currently fixing some race conditions but I have peers joining (haven't pushed yet)

File: src/lib/trust_system/record.ml [hunk 1]
```diff
@@ -34,7 +34,7 @@ end) : S = struct
 
   let bans_disabled =
     let module Expiration = struct
-      [%%expires_after "20190907"]
+      [%%expires_after "20190914"]
```
Comment (bkase) [inline src/lib/trust_system/record.ml]: one more week?

File: src/lib/gossip_net/gossip_net.ml [hunk 2]
```diff
@@ -639,9 +647,13 @@ module Make (Message : Message_intf) : S with type msg := Message.msg = struct
...
-                        let pushback = Pipe.write w
-                          ( Node_addrs_and_ports.to_peer config.addrs_and_ports
-                          |> Peer.to_yojson |> Yojson.Safe.to_string ) in Pipe.close w ; pushback )
+                        let pushback =
+                          Pipe.write w
+                            ( Node_addrs_and_ports.to_peer
+                                config.addrs_and_ports
+                            |> Peer.to_yojson |> Yojson.Safe.to_string )
+                        in
+                        Pipe.close w ; pushback )
```
Comment (bkase) [inline src/lib/gossip_net/gossip_net.ml]: is this close safe? Does the close need to happen in `Deferred.upon` of the pushback?

File: src/lib/gossip_net/gossip_net.ml [hunk 3]
```diff
@@ -565,6 +541,39 @@ module Make (Message : Message_intf) : S with type msg := Message.msg = struct
               failwithf "Unhandled Membership.connect exception: %s"
                 (Exn.to_string e) ()
         in
+        let peer_event_reader, peer_event_writer = Linear_pipe.create () in
+        Linear_pipe.iter_unordered peer_event_reader ~max_concurrency:64
+          ~f:(fun event ->
+            let%bind t = t_for_restarting in
```
Comment (bkase) [inline src/lib/gossip_net/gossip_net.ml]: this doesn't actually support restarting now right?

File: src/lib/gossip_net/gossip_net.ml (hunk 2 above)
Comment (cmr) [reply src/lib/gossip_net/gossip_net.ml]: This is safe: it synchronously prevents new writes from entering the queue, without modifying the contents of the queue.

File: src/lib/gossip_net/gossip_net.ml (hunk 3 above)
Comment (cmr) [reply src/lib/gossip_net/gossip_net.ml]: Restarting should continue to work the same as it has with this PR. The libp2p stuff doesn't support being restarted yet, but the old Haskell stuff does.

File: src/app/cli/src/init/client.ml [hunk 4]
```diff
@@ -1105,19 +1105,22 @@ let generate_libp2p_keypair =
        @@ fun () ->
        Deferred.ignore
          (let open Deferred.Let_syntax in
-         let logger = Logger.create () in
+         (* FIXME: I'd like to accumulate messages into this logger and only dump them out in failure paths. *)
```
Comment (nholland94) [inline src/app/cli/src/init/client.ml]: Can we open an issue to add this feature to the logger?

Comment (cmr) [reply src/app/cli/src/init/client.ml]: yes.

**Pull request merged**